File: src/main/java/org/hisp/dhis/rules/models/RuleActionMessage.java
package org.hisp.dhis.rules.models;

import javax.annotation.CheckForNull;
import javax.annotation.Nonnull;

import static org.hisp.dhis.rules.models.AttributeType.UNKNOWN;

public record RuleActionMessage(
        @Nonnull String data,
        @Nonnull AttributeType attributeType,
        @Nonnull String content,
        @Nonnull String field,
        @Nonnull Type type
) implements RuleActionAttribute {

    public enum Type { WARNING_ON_COMPILATION, ERROR_ON_COMPILATION, SHOW_WARNING, SHOW_ERROR }

    /*
    On Completion
     */
    @Nonnull
    public static RuleActionMessage create( @CheckForNull String content,
                                                      @CheckForNull String data, @CheckForNull String field, @CheckForNull AttributeType attributeType, @Nonnull Type type )
    {
        if ( content == null && data == null && field == null )
        {
            throw new IllegalArgumentException( "Content, data and field" +
                    " must not be null at the same time" );
        }

        return new RuleActionMessage( data == null ? "" : data, attributeType == null ? UNKNOWN : attributeType,
                content == null ? "" : content, field == null ? "" : field, type );
    }

    @Nonnull
    public static RuleActionMessage create(@CheckForNull String content, @CheckForNull String data, @CheckForNull String field, @Nonnull Type type)
    {
        return create( content, data, field, UNKNOWN, type );
    }
}

File: src/main/java/org/hisp/dhis/rules/models/RuleActionMessage.kt
package org.hisp.dhis.rules.models

import javax.annotation.Nonnull

data class RuleActionMessage(
    val type: Type,
    val field: String = "",
    val content: String = "",
    val attributeType: AttributeType = AttributeType.UNKNOWN,
    val data: String = ""
) : RuleActionAttribute {
    enum class Type {
        WARNING_ON_COMPILATION,
        ERROR_ON_COMPILATION,
        SHOW_WARNING,
        SHOW_ERROR
    }

    companion object {
        /*
    On Completion
     */
        fun create(
            content: String?,
            data: String?, field: String?, attributeType: AttributeType?, type: Type
        ): RuleActionMessage {
            require(!(content == null && data == null && field == null)) {
                "Content, data and field" +
                        " must not be null at the same time"
            }
            return RuleActionMessage(
                type,field ?: "",  content ?: "",
                attributeType ?: AttributeType.UNKNOWN, data ?: ""
            )
        }
    }

    override fun attributeType(): AttributeType {
        return attributeType
    }

    override fun data(): String {
        return data
    }
}
